### PR TITLE
Phi-2

### PR DESCRIPTION
**Pull request by matheper**
This PR contains a set of small improvements and adaptations made for experiments using Phi-2 model.
Most of this code was previously reviewed and merged into the `phi2` branch.
Now it is being merged into the `main` branch.

- Loading model configs from yaml
- Fix a hard-coded tokenization cleaning in dln.score
- Mini-batch calls to VLLM models
- Check and fail early if a request is larger than the model's maximum sequence length
- v3.6 backward template
- Add top-k to read_results
- Add new scripts
- Document the DLN components in the main README
- Document and type hinting dln.datasets

TODO in a separate PR:
- Organize scripts in a better way / separate folders
- Document OpenAI change not allowing to request `echo` and `logprobs` at the same time

### CODE REVIEW

**Review by chisingh — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -50,9 +50,200 @@ Then, export the openai env vars `OPENAI_API_BASE` and `OPENAI_API_KEY`. Remembe
...
+fwd_model = llm_registry.register(
+    "fwd_model",  # how you refer to the model
+    "gpt-35-turbo-instruct",  # model id
+    temperature=0.0,
+    max_tokens=256,
+    stop=None,
+)
+
+bwd_model = llm_registry.register(
+    "fwd_model",
```
Comment (MarcCote) [inline README.md:113]:
```suggestion
    "bwd_model",
```

File: README.md [hunk 2]
```diff
@@ -50,9 +50,200 @@ Then, export the openai env vars `OPENAI_API_BASE` and `OPENAI_API_KEY`. Remembe
...
+    "gpt-35-turbo-instruct",
+    temperature=0.7,
+    max_tokens=512,
+    stop=None,
+)
+
+fwd_model.generate("What is putine?")
```
Comment (MarcCote) [inline README.md:120]:
```suggestion
fwd_model.generate("What is sirop d'érable?")
```

File: README.md [hunk 3]
```diff
@@ -50,9 +50,200 @@ Then, export the openai env vars `OPENAI_API_BASE` and `OPENAI_API_KEY`. Remembe
...
+fwd_model.generate("What is putine?", max_tokens=200, stop=[r"\n"])
```
Comment (MarcCote) [inline README.md:121]:
```suggestion
fwd_model.generate("What is sirop d'érable?", max_tokens=200, stop=[r"\n"])
```

File: README.md [hunk 4]
```diff
@@ -50,9 +50,200 @@ Then, export the openai env vars `OPENAI_API_BASE` and `OPENAI_API_KEY`. Remembe
...
+
+```python
+from dln.loss import LossRegistry
+from dln.postprocessing import postprocess_prediction
+
+LossRegistry.available_losses()  # list available losses
+loss_fn = LossRegistry.instantiate(
+    "exact_match_loss", postprocess_prediction
+)
+losses = loss_fn(y_hat, y)
```
Comment (MarcCote) [inline README.md:192]: `y_hat` and `y` are not defined.

File: README.md [hunk 5]
```diff
@@ -50,9 +50,200 @@ Then, export the openai env vars `OPENAI_API_BASE` and `OPENAI_API_KEY`. Remembe
...
+    next_prompt="What is the capital of this country",
+    num_samples=10,
+)
+
+score_request = ScoreRequest(
+    context="What is the capital of this country: Canada",
+    target="Ottawa",
+    payload="Ottawa",
+)
+score = logprobs_score.score_requests([score_request])
```
Comment (MarcCote) [inline README.md:236]: Got this error when running this but I see you have listed it as a todo.
```
InvalidRequestError: Setting 'echo' and 'logprobs' at the same time is not supported for this model.
```

File: README.md [hunk 6]
```diff
@@ -81,7 +272,7 @@ If you find DLNs useful, please consider citing this work!
 
 ```text
 @article{sordoni2023deep,
-      title={Deep Language Networks: Joint Prompt Training of Stacked LLMs using Variational Inference}, 
+      title={Deep Language Networks: Joint Prompt Training of Stacked LLMs using Variational Inference},
```
Comment (MarcCote) [inline README.md:275]:
```suggestion
      title={Joint Prompt Optimization of Stacked LLMs using Variational Inference},
```

File: dln/dataset.py [hunk 7]
```diff
@@ -344,6 +395,36 @@ def init_dataset(
     return dataset
 
 
+def option_shuffle(data_point, rng):
+    import re
```
Comment (MarcCote) [inline dln/dataset.py:399]: Do we want to move this import at the top of the file?

File: projects/vi_dln/read_results.py [hunk 8]
```diff
@@ -67,21 +65,70 @@ def mean_confidence_interval(data, confidence=0.95):
...
+    top_k_argmax_mean_test, _ = top_k(v, len(v["test"]))
+
+    # TODO: receive min_seeds as parameter
+    if len(v["dev"]) <= 2:
+        continue
     data.append(
         {
-            "name": k,
+            # "name": k,
+            # "name": k[:30],
```
Comment (MarcCote) [inline projects/vi_dln/read_results.py:115]:
```suggestion
```

File: projects/vi_dln/scripts/one_layer/phi2_DLN-1.sh [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+posterior_temp=1.
+p_hidden_tpl="suffix_forward_tbs"
+q_hidden_tpl="suffix_forward_tbs_y|suffix_forward_tbs"
+fwd_max_tokens=256
+bwd_max_tokens=512
+
+#iters=20
+#batch_size=20
+#q_prompt_tpl="q_action_prompt:v3.5"
+#logp_penalty=2
```
Comment (MarcCote) [inline projects/vi_dln/scripts/one_layer/phi2_DLN-1.sh:31]: Do we want to remove this?

File: projects/vi_dln/scripts/one_layer/phi2_DLN-1.sh [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+            --q_prompt ${q_prompt_tpl} \
+            --logp_penalty ${logp_penalty} \
+            --posterior_temp ${posterior_temp} \
+            --p_hidden ${p_hidden_tpl} \
+            --q_hidden ${q_hidden_tpl} \
+            --seed ${seed} \
+            --fwd_max_tokens ${fwd_max_tokens} \
+            --bwd_max_tokens ${bwd_max_tokens} \
+            --output_scoring_function ${output_scoring_function}
+            # --balance_batch \
```
Comment (MarcCote) [inline projects/vi_dln/scripts/one_layer/phi2_DLN-1.sh:85]: Remove this as well?

File: projects/vi_dln/scripts/one_layer/phi2_Nshot.sh [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+logp_penalty=1.
+posterior_temp=1.
+p_hidden_tpl="suffix_forward_tbs"
+q_hidden_tpl="suffix_forward_tbs_y|suffix_forward_tbs"
+fwd_max_tokens=256
+bwd_max_tokens=512
+one_layer=True
+
+# iters=20
+# num_p_samples=15
```
Comment (MarcCote) [inline projects/vi_dln/scripts/one_layer/phi2_Nshot.sh:32]: Remove?

File: README.md (hunk 5 above)
Comment (matheper) [reply README.md:236]: Actually, I will do it in this PR. Somewhere in the readme add a warning mentioning that people can encounter this error if using openai models.

Comment (matheper) [reply README.md:245]: Added warning message to main and vi README files

**Comment by matheper** [comment]
Thanks for re-reviewing everything @MarcCote!

File: projects/vi_dln/read_results.py [hunk 12]
```diff
@@ -95,7 +95,7 @@ def top_k(data, k):
     return mean_test_score, std_test_score
 
 def find_dataset_name(dataset_path):
-    dataset_names = ["navigate", "hyperbaton", "date_understanding", "logical_deduction_seven_objects", "disaster", "airline", "mpqa", "trec", "subj"]
+    dataset_names = ["gsm8k", "navigate", "hyperbaton", "date_understanding", "logical_deduction_seven_objects", "disaster", "airline", "mpqa", "trec", "subj"]
```
Comment (MarcCote) [inline projects/vi_dln/read_results.py:98]: Good catch

**Review by MarcCote — APPROVED** [review]

**Pull request merged**